test/tlsconnection.js

```diff
@@ -0,0 +1,1160 @@
+      const recordContent = new BufferWriter();
+      recordContent.writeBytes(await testHelpers.makeEncryptedExtensionsMessage({}));
+      recordContent.writeBytes(new Finished(zeros(32)).toBytes());
+      const encExtAndFinished = await testHelpers.makeEncryptedRecord(cipherstate, {
+        content: recordContent.flush(),
+        type: 22,
+      });
+      await assert.throwsAsync(async () => {
+        await client.recv(encExtAndFinished);
+      }, TLSError, 'DECRYPT_ERROR');
+    });
+
+    it('errors if it receives a non-empty EncryptedExtensions', async () => {
+      const cipherstate = await testHelpers.makeCipherState(client._keyschedule.serverHandshakeTrafficSecret, 0);
+      const encExt = await testHelpers.makeEncryptedRecord(cipherstate, {
+        content: await testHelpers.makeEncryptedExtensionsMessage({
+          extensions: [testHelpers.makeSupportedVersionsExtension([0x0304])]
+        }),
+        type: 22,
+      });
+      await assert.throwsAsync(async () => {
+        await client.recv(encExt);
+      }, TLSError, 'UNSUPPORTED_EXTENSION');
+    });
+  });
+
+  it('accepts a ServerHello with extensions listed in a different order', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makePreSharedKeyExtension(0),
+          testHelpers.makeSupportedVersionsExtension(0x0304)
+        ]
+      }),
+      type: 22
+    });
+    const data = await client.recv(serverHello);
+    assert.equal(data, null);
+  });
+
+  it('errors if it receives non-TLS-handshake-message data', async () => {
+    await assert.throwsAsync(async () => {
+      await client.recv(utf8ToBytes('a badly-formed message'));
+    }, TLSError, 'DECODE_ERROR');
+  });
+
+  it('errors if it receives a message of unknown type', async () => {
+    await assert.throwsAsync(async () => {
+      await client.recv(hexToBytes('0103030002AABB'));
+    }, TLSError, 'UNEXPECTED_MESSAGE');
+  });
+
+  it('errors if it receives a message other than ServerHello', async () => {
+    await assert.throwsAsync(async () => {
+      await client.recv(TEST_VECTORS.CLIENT_HELLO);
+    }, TLSError, 'UNEXPECTED_MESSAGE');
+  });
+
+  it('errors if the version number in the ServerHello predates TLS1.2', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        version: 0x0123
+      }),
+      type: 22
+    });
+    await assert.throwsAsync(async () => {
+      await client.recv(serverHello);
+    }, TLSError, 'ILLEGAL_PARAMETER');
+  });
+
+  it('errors if the ServerHello did not send supported_versions', async () => {
+    const serverHello = testHelpers.makePlaintextRecord({
+      content: await testHelpers.makeServerHelloMessage({
+        extensions: [
+          testHelpers.makePreSharedKeyExtension(1)
+        ]
```

Comment:
Maybe add version: 0x0304 to this, maybe as a supplementary test, so that you catch the case where the version field in the SH is used to drive version negotiation.  (Not that your code is at all vulnerable to this currently, but it's a good paranoia check.)

Reply:
To be clear, the handshake should *fail* if `SH.version` is `0x0304`, right?

Reply:
Correct.  Fail.